src/utils/data-access.ts

```diff
@@ -0,0 +1,148 @@
+  //     DataAccess.instance = new DataAccess();
+  //   }
+
+  //   return DataAccess.instance;
+  // }
+
+  private getById(options: {
+    indexName: string;
+    keyCondition: string;
+    expressionAttributeValues: { [key: string]: string };
+  }): Promise<PromiseResult<QueryOutput, AWSError>> {
+    const params = {
+      TableName: this.tableName,
+      IndexName: options.indexName,
+      KeyConditionExpression: options.keyCondition,
+      ExpressionAttributeValues: options.expressionAttributeValues,
+    };
+    debug('params getbyId', params);
+    return this.docClient.query(params).promise();
+  }
+
+  private async put<T>(
+    payload: T,
+    options?: { conditionalExpression: string; expressionAttributeValues: { [key: string]: string } },
+  ): Promise<PromiseResult<DocumentClient.PutItemOutput, AWSError>> {
+    let query = {
+      TableName: this.tableName,
+      Item: payload,
+      ConditionExpression: undefined,
+      ExpressionAttributeValues: undefined,
+    };
+    if (options) {
+      query = {
+        ...query,
+        ConditionExpression: options.conditionalExpression,
+        ExpressionAttributeValues: options.expressionAttributeValues,
+      };
+    }
+    return this.docClient.put(query).promise();
+  }
+
+  public async upsertTrailerRegistration(
+    trailerRegistration: domain.TrailerRegistration,
+  ): Promise<domain.TrailerRegistration> {
+    await this.put<domain.TrailerRegistration>(trailerRegistration);
+    const result = await this.getByVinOrChassisWithMake(trailerRegistration.vinOrChassisWithMake);
+    if (result && result.length) {
+      return result[0];
+    }
+    return null;
+  }
+
+  public async getByVinOrChassisWithMake(vinOrChassisWithMake: string): Promise<domain.TrailerRegistration[]> {
+    const getParams = {
+      indexName: 'vinOrChassisWithMakeIndex',
+      keyCondition: 'vinOrChassisWithMake = :vinOrChassisWithMake',
+      expressionAttributeValues: {
+        ':vinOrChassisWithMake': vinOrChassisWithMake,
+      },
+    };
+    const data = (await this.getById(getParams)) as { Count: number; Items: unknown };
+    if (!data || !data.Count) {
+      return null;
+    }
+    return data.Items as domain.TrailerRegistration[];
+  }
+
+  public async getByTrn(trn: string): Promise<domain.TrailerRegistration[]> {
+    const getParams = {
+      indexName: 'trnIndex',
+      keyCondition: 'trn = :trn',
+      expressionAttributeValues: {
+        ':trn': trn,
+      },
+    };
+    const data = (await this.getById(getParams)) as { Count: number; Items: unknown };
+    if (!data || !data.Count) {
+      return null;
+    }
+    return data.Items as domain.TrailerRegistration[];
+  }
+
+  public createMultiple(
+    trailerRegistrations: domain.TrailerRegistration[],
+  ): Promise<PromiseResult<DocumentClient.BatchWriteItemOutput, AWSError>> {
+    const params = this.generateBatchWritePartialParams();
+
+    trailerRegistrations.forEach((trailerRegistration) => {
```

Comment:
Probably too early to optimise but is there a limit in terms of data we can process on batch? Can we get in situations where we may have to process lots of records @bristol-mike-jones

Reply:
This is used only for integration tests. The seed file only has 4 values. We have more complex data in test-results and tech-records which are doing similar behavior on integration tests and so far has been working without problems.